design-proposals/vdpa-interfaces.md

```diff
@@ -0,0 +1,146 @@
+# Overview
+A vDPA device means a type of device whose datapath complies with the virtio specification, but whose control path is vendor specific. Currently, two vDPA drivers are implemented:
+- vhost interface (vhost-vdpa) for userspace or guest virtio driver, like a VM running in QEMU
+- virtio interface (virtio-vdpa) for bare-metal or containerized applications running in the host
+
+This design provides an approach for creating a VM with vhost-vdpa interface.
+
+## Goals
+Have a mechanism to allow creation of virtual machines with vdpa interfaces.
+
+## Non Goals
+Because the implementation of vDPA can be SR-IOV based or not, this mechanism should not be coupled with the [Intel SR-IOV device plugin](https://github.com/intel/sriov-network-device-plugin) and [SR-IOV CNI plugin](https://github.com/k8snetworkplumbingwg/sriov-cni).
+
+## User Stories
+As a Kubevirt user I would like to create a vm with vdpa interfaces.
+
+## Repos
+[KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+Design can be inspired by the existing [sriov type interface](https://kubevirt.io/user-guide/virtual_machines/interfaces_and_networks/#sriov), but you can see the difference by comparing the libvirt domain XML:
+```xml
+<hostdev type="pci" managed="no" mode="subsystem">
+        <source>
+                <address type="" domain="0x0000" bus="0xcc" slot="0x00" function="0x4"></address>
+        </source>
+        <address type="pci" domain="0x0000" bus="0x07" slot="0x00" function="0x0"></address>
+        <alias name="ua-sriov-offload-ovn"></alias>
+</hostdev>
+```
+```xml
+<interface type="vdpa">
+        <source dev="/dev/vhost-vdpa-1"></source>
+        <model type="virtio"></model>
+        <alias name="ua-offload-ovn"></alias>
+        <driver name="vhost" queues="4"></driver>
+</interface>
+```
+- The former is hostdev while the latter is interface
+- The source of the former is a bdf while the source of the latter is a device path.
+
+## API
+Add a vDPA type to InterfaceBindingMethod:
+```go
+// Represents the method which will be used to connect the interface to the guest.
+// Only one of its members may be specified.
+type InterfaceBindingMethod struct {
+	Bridge     *InterfaceBridge     `json:"bridge,omitempty"`
+	Slirp      *InterfaceSlirp      `json:"slirp,omitempty"`
+	Masquerade *InterfaceMasquerade `json:"masquerade,omitempty"`
+	SRIOV      *InterfaceSRIOV      `json:"sriov,omitempty"`
+	VDPA       *InterfaceVDPA       `json:"vdpa,omitempty"`
```

Comment:
I do wonder if we can avoid this.

Couldn't we rather work on a "protocl" to let a device plugin "expose" what kind of binding it needs?

Aka internally in kubevirt we enable vDPA, but there is no VM API flag for this. Instead the CNI is signaling (can it set an env var?) to request vDPA. KubeVirt will meet this request if possible.

Reply:
FWIW - This could then be exteneded to other bindings as well, eventually making it optional for some cnis and bindings.

Reply:
@AlonaKaplan @EdDev ?

Reply:
@fabiand lmillei is working on a vdpa binding plugin. He added a reference to it in [this](https://github.com/kubevirt/community/pull/228#issuecomment-1997241806) comment.

Reply:
Thanks, I understand: But as this is a proposal for KubeVirt I wnated to check with you as a chair of sig-network what you think about such a new internal API

Reply:
I don't see any reason to add it as an internal API for VDPA. The network binding plugin should fit the use case. We are now working on extending the binding plugin API to have downward API so the env vars/device-info will get to the sidecar building the libvirt domain.
1. [Device info](https://docs.google.com/document/d/13fLaqZuM8El_CT5XzxPv-1epKsB1-Spv_jtXveEkkdQ/edit#heading=h.ra95t07p2hox). 
2. vdpa [concrete](https://docs.google.com/document/d/13fLaqZuM8El_CT5XzxPv-1epKsB1-Spv_jtXveEkkdQ/edit#heading=h.vgfhk652zjg) example.

> Couldn't we rather work on a "protocl" to let a device plugin "expose" what kind of binding it needs?
> Aka internally in kubevirt we enable vDPA, but there is no VM API flag for this. Instead the CNI is signaling (can it set an env var?) to request vDPA. KubeVirt will meet this request if possible.

Regarding this suggestion, I"m not sure I understand it.
What do you mean by internally in kubevirt enable vDPA? kuevbirt core code would have to be familiar with vDPA but will have to discover that this is the binding it should use based on the information from the device plugin?
Why do we need this third approach? We already have core bindings like masquerade, bridge and SRIOV and we have the binding plugin mechanism for other bindings.